File: lib/go/protochain/solana/program/token/v1/metadata.go
// Package token_v1 provides off-chain metadata types for Metaplex Token Metadata.
//
// Covers all five token standards:
//   - 0 – NonFungible
//   - 1 – FungibleAsset
//   - 2 – Fungible
//   - 3 – NonFungibleEdition
//   - 4 – ProgrammableNonFungible
//
// See: https://www.metaplex.com/docs/smart-contracts/token-metadata/token-standard
package token_v1

import (
	"encoding/json"
	"fmt"
)

// MetadataAttributeDisplayType controls how wallets / marketplaces render a
// numeric attribute. An empty string means the default string rendering.
//
// See: https://docs.opensea.io/docs/metadata-standards
type MetadataAttributeDisplayType string

const (
	DisplayTypeNumber          MetadataAttributeDisplayType = "number"
	DisplayTypeBoostNumber     MetadataAttributeDisplayType = "boost_number"
	DisplayTypeBoostPercentage MetadataAttributeDisplayType = "boost_percentage"
	DisplayTypeDate            MetadataAttributeDisplayType = "date"
)

// MetadataCategory describes the primary media type of an asset.
type MetadataCategory string

const (
	CategoryImage MetadataCategory = "image"
	CategoryVideo MetadataCategory = "video"
	CategoryAudio MetadataCategory = "audio"
	CategoryVR    MetadataCategory = "vr"
	CategoryHTML  MetadataCategory = "html"
)

// AttributeValue holds the polymorphic value of a MetadataAttribute.
// The underlying type is one of: string, float64, or bool.
type AttributeValue struct {
	raw any
}

// StringValue returns the value as a string and true, or ("", false).
func (v AttributeValue) StringValue() (string, bool) {
	s, ok := v.raw.(string)
	return s, ok
}

// Float64Value returns the value as a float64 and true, or (0, false).
func (v AttributeValue) Float64Value() (float64, bool) {
	f, ok := v.raw.(float64)
	return f, ok
}

// BoolValue returns the value as a bool and true, or (false, false).
func (v AttributeValue) BoolValue() (bool, bool) {
	b, ok := v.raw.(bool)
	return b, ok
}

// Raw returns the underlying value (string | float64 | bool).
func (v AttributeValue) Raw() any { return v.raw }

// String implements fmt.Stringer.
func (v AttributeValue) String() string { return fmt.Sprintf("%v", v.raw) }

// MarshalJSON implements json.Marshaler.
func (v AttributeValue) MarshalJSON() ([]byte, error) {
	return json.Marshal(v.raw)
}

// UnmarshalJSON implements json.Unmarshaler.
// Accepts JSON strings, numbers, and booleans.
func (v *AttributeValue) UnmarshalJSON(data []byte) error {
	var raw any
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}
	switch raw.(type) {
	case string, float64, bool:
		v.raw = raw
	case nil:
		v.raw = ""
	default:
		return fmt.Errorf("unsupported attribute value type: %T", raw)
	}
	return nil
}

// MetadataAttribute represents a single trait / attribute of an asset.
//
// Value is intentionally polymorphic:
//   - string  → rendered as a tag / pill
//   - float64 → rendered as bar / numeric (depends on DisplayType)
//   - bool    → rendered as a toggle / checkbox
//
// When DisplayType is "date", Value must be a Unix timestamp (seconds).
type MetadataAttribute struct {
	TraitType string         `json:"trait_type"`
	Value     AttributeValue `json:"value"`
	// DisplayType controls how wallets / marketplaces render the value.
	DisplayType MetadataAttributeDisplayType `json:"display_type,omitempty"`
	// MaxValue is the upper bound for numeric traits (progress bars).
	MaxValue *float64 `json:"max_value,omitempty"`
}

// MetadataFile is a file entry in the asset's file manifest.
type MetadataFile struct {
	// URI is the fully-qualified location (IPFS, Arweave, HTTPS).
	URI string `json:"uri"`
	// Type is the MIME type (e.g. "image/png", "video/mp4").
	Type string `json:"type"`
	// CDN indicates the URI is served through a Content Delivery Network.
	CDN *bool `json:"cdn,omitempty"`
}

// MetadataCreator is the deprecated off-chain creator entry inside
// properties.creators. Canonical creator / royalty data lives on-chain
// in the Metadata account.
type MetadataCreator struct {
	// Address is the Base-58 encoded public key of the creator.
	Address string `json:"address"`
	// Share is the percentage of royalties (0-100, must sum to 100).
	Share int `json:"share"`
}

// MetadataProperties holds the properties bag attached to FungibleAsset,
// NonFungible, and ProgrammableNonFungible metadata.
type MetadataProperties struct {
	// Files is the multi-file manifest for the asset's media.
	Files []MetadataFile `json:"files,omitempty"`
	// Category is the primary media category.
	Category MetadataCategory `json:"category,omitempty"`
	// Deprecated: Creators — on-chain creators are authoritative.
	// Retained for backwards-compatibility with existing off-chain JSON.
	Creators []MetadataCreator `json:"creators,omitempty"`
}

// MetadataCollection is the deprecated off-chain collection grouping.
// Canonical collection data lives on-chain via the Collection field in
// the Metadata account.
type MetadataCollection struct {
	// Name is the human-readable name of the collection.
	Name string `json:"name,omitempty"`
	// Family is the broad family grouping (e.g. "Solana Monkey Business").
	Family string `json:"family,omitempty"`
}

// UniversalTokenMetadata is the universal Metaplex off-chain metadata schema.
//
// It is a superset of all five token standards:
//   - Fungible uses: Name, Symbol, Description, Image
//   - FungibleAsset / NonFungible / NonFungibleEdition / ProgrammableNonFungible
//     add: AnimationURL, ExternalURL, Attributes, Properties, and more
//
// Fields marked deprecated are retained because real-world JSON files still
// contain them. Canonical data for those fields lives on-chain.
//
// See: https://www.metaplex.com/docs/smart-contracts/token-metadata/token-standard
type UniversalTokenMetadata struct {
	// Name is the display name of the asset.
	Name string `json:"name"`
	// Symbol is the ticker symbol (e.g. "USDC"). Required for Fungible tokens.
	Symbol string `json:"symbol,omitempty"`
	// Description is the human-readable description. Supports markdown.
	Description string `json:"description"`
	// Image is the URI to the asset's primary image.
	// For NFTs with an AnimationURL this acts as the thumbnail / poster.
	Image string `json:"image"`
	// AnimationURL is the URI to the rich-media asset (mp4, glb, html, mp3, etc.).
	AnimationURL string `json:"animation_url,omitempty"`
	// ExternalURL is a link to an external page for the project / issuer.
	ExternalURL string `json:"external_url,omitempty"`
	// BackgroundColor is a six-character hex colour (no "#" prefix) used as the
	// background behind the image on marketplaces like OpenSea.
	BackgroundColor string `json:"background_color,omitempty"`
	// Attributes is the array of traits used for rarity, filtering, and display.
	Attributes []MetadataAttribute `json:"attributes,omitempty"`
	// Properties holds the file manifest, media category, and deprecated creators.
	Properties *MetadataProperties `json:"properties,omitempty"`
	// Deprecated: Collection — on-chain Collection field is authoritative.
	// Retained for backwards-compatibility.
	Collection *MetadataCollection `json:"collection,omitempty"`
	// Deprecated: SellerFeeBasisPoints — royalties in basis points (100 = 1%).
	// Canonical value lives on-chain. Retained because many existing JSON files
	// still include it.
	SellerFeeBasisPoints *int `json:"seller_fee_basis_points,omitempty"`
}

// ParseUniversalTokenMetadata unmarshals raw JSON bytes into a
// UniversalTokenMetadata. It returns an error if the JSON is malformed.
func ParseUniversalTokenMetadata(data []byte) (*UniversalTokenMetadata, error) {
	var m UniversalTokenMetadata
	if err := json.Unmarshal(data, &m); err != nil {
		return nil, fmt.Errorf("parsing token metadata: %w", err)
	}
	return &m, nil
}

File: lib/ts-web/src/protochain/solana/program/token/v1/metadata.ts
/**
 * Off-chain metadata types for Metaplex Token Metadata.
 *
 * Covers all five token standards:
 *   0 – NonFungible
 *   1 – FungibleAsset
 *   2 – Fungible
 *   3 – NonFungibleEdition
 *   4 – ProgrammableNonFungible
 *
 * @see {@link https://www.metaplex.com/docs/smart-contracts/token-metadata/token-standard}
 */

// ---------------------------------------------------------------------------
// Attributes
// ---------------------------------------------------------------------------

/**
 * Display types that instruct wallets / marketplaces how to render a numeric
 * attribute. Omitting display_type defaults to a plain string trait.
 *
 * @see {@link https://docs.opensea.io/docs/metadata-standards}
 */
export type MetadataAttributeDisplayType = 'number' | 'boost_number' | 'boost_percentage' | 'date';

/**
 * A single trait / attribute of an asset.
 *
 * `value` is intentionally broad:
 *   - string  → rendered as a tag / pill
 *   - number  → rendered as bar / numeric (depends on display_type)
 *   - boolean → rendered as a toggle / checkbox
 *
 * When `display_type` is `"date"`, `value` must be a Unix timestamp (seconds).
 */
export interface MetadataAttribute {
  trait_type: string;
  value: string | number | boolean;
  /** Controls how wallets / marketplaces render the value. */
  display_type?: MetadataAttributeDisplayType;
  /** Upper bound for numeric traits (used by marketplaces for progress bars). */
  max_value?: number;
}

// ---------------------------------------------------------------------------
// Files & creators inside `properties`
// ---------------------------------------------------------------------------

/** Known media categories for the `properties.category` field. */
export type MetadataCategory = 'image' | 'video' | 'audio' | 'vr' | 'html';

/**
 * A file entry in the asset's file manifest.
 *
 * The manifest lets wallets select the best representation
 * (e.g. thumbnail vs. full-res, video vs. still).
 */
export interface MetadataFile {
  /** Fully-qualified URI (IPFS, Arweave, HTTPS). */
  uri: string;
  /** MIME type (e.g. "image/png", "video/mp4", "model/gltf-binary"). */
  type: string;
  /** If true, the URI is served through a CDN. */
  cdn?: boolean;
}

/**
 * Off-chain creator entry inside `properties.creators`.
 *
 * NOTE: This is the **deprecated** off-chain representation.
 * Canonical creator / royalty data lives on-chain in the Metadata account.
 * Included here because many existing assets still carry it in their JSON.
 */
export interface MetadataCreator {
  /** Base-58 encoded public key of the creator. */
  address: string;
  /** Percentage share of royalties (0-100, must sum to 100 across creators). */
  share: number;
}

/**
 * The `properties` bag attached to FungibleAsset, NonFungible, and
 * ProgrammableNonFungible metadata.
 */
export interface MetadataProperties {
  /** Multi-file manifest for the asset's media. */
  files?: MetadataFile[];
  /** Primary media category – drives how wallets render the asset. */
  category?: MetadataCategory;
  /**
   * @deprecated On-chain creators are authoritative. Included for
   * backwards-compatibility with existing off-chain JSON.
   */
  creators?: MetadataCreator[];
}

// ---------------------------------------------------------------------------
// Collection (off-chain, deprecated)
// ---------------------------------------------------------------------------

/**
 * Off-chain collection grouping.
 *
 * NOTE: This is the **deprecated** off-chain convention. Canonical collection
 * data lives on-chain via the Collection field in the Metadata account.
 * Included here because many existing assets still carry it in their JSON.
 */
export interface MetadataCollection {
  /** Human-readable name of the collection. */
  name?: string;
  /** Broad family grouping (e.g. "Solana Monkey Business"). */
  family?: string;
}

// ---------------------------------------------------------------------------
// Universal metadata schema
// ---------------------------------------------------------------------------

/**
 * The universal Metaplex off-chain metadata schema.
 *
 * Designed to be a superset of all five token standards:
 *   - **Fungible** uses: name, symbol, description, image
 *   - **FungibleAsset / NonFungible / NonFungibleEdition / ProgrammableNonFungible**
 *     add: animation_url, external_url, attributes, properties, and more
 *
 * Fields marked deprecated are retained because real-world JSON files still
 * contain them. Canonical data for those fields lives on-chain.
 *
 * @see {@link https://www.metaplex.com/docs/smart-contracts/token-metadata/token-standard}
 */
export interface UniversalTokenMetadata {
  /** The display name of the asset. */
  name: string;

  /** Ticker symbol (e.g. "USDC"). Required for Fungible tokens. */
  symbol?: string;

  /** Human-readable description. Supports markdown. */
  description: string;

  /**
   * URI to the asset's primary image.
   * For NFTs with an animation_url this acts as the thumbnail / poster.
   */
  image: string;

  /** URI to the rich-media asset (mp4, glb, html, mp3, etc.). */
  animation_url?: string;

  /** URL to an external page for the project / issuer. */
  external_url?: string;

  /**
   * Six-character hex colour (no "#" prefix) used as the background
   * behind the image on marketplaces like OpenSea.
   */
  background_color?: string;

  /** Array of traits used for rarity, filtering, and display. */
  attributes?: MetadataAttribute[];

  /** File manifest, media category, and (deprecated) creator list. */
  properties?: MetadataProperties;

  /**
   * @deprecated Off-chain collection grouping. On-chain Collection field is
   * authoritative. Retained for backwards-compatibility.
   */
  collection?: MetadataCollection;

  /**
   * @deprecated Royalties in basis points (100 = 1%). Canonical value lives
   * on-chain in the Metadata account's `seller_fee_basis_points` field.
   * Retained because many existing JSON files still include it.
   */
  seller_fee_basis_points?: number;
}

// ---------------------------------------------------------------------------
// Parser
// ---------------------------------------------------------------------------

export class MetadataParser {
  /**
   * Parses raw JSON into a typed {@link UniversalTokenMetadata} object.
   *
   * Performs defensive type-coercion and null-coalescing so that malformed
   * third-party metadata does not cause runtime crashes.
   *
   * @param json - The raw JSON object fetched from a Token Metadata URI.
   * @returns A sanitised UniversalTokenMetadata.
   */
  static fromJSON(json: unknown): UniversalTokenMetadata {
    if (json == null || typeof json !== 'object') {
      throw new Error('MetadataParser: input must be a non-null object.');
    }

    const j = json as Record<string, unknown>;

    return {
      name: asString(j.name, ''),
      symbol: optString(j.symbol),
      description: asString(j.description, ''),
      image: asString(j.image, ''),
      animation_url: optString(j.animation_url),
      external_url: optString(j.external_url),
      background_color: optString(j.background_color),

      attributes: parseAttributes(j.attributes),
      properties: parseProperties(j.properties),
      collection: parseCollection(j.collection),
      seller_fee_basis_points: optNumber(j.seller_fee_basis_points),
    };
  }
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function asString(v: unknown, fallback: string): string {
  return v != null ? String(v) : fallback;
}

function optString(v: unknown): string | undefined {
  return v != null ? String(v) : undefined;
}

function optNumber(v: unknown): number | undefined {
  if (v == null) return undefined;
  const n = Number(v);
  return Number.isFinite(n) ? n : undefined;
}

function parseAttributes(v: unknown): MetadataAttribute[] | undefined {
  if (!Array.isArray(v)) return undefined;

  return v
    .filter((a): a is Record<string, unknown> => a != null && typeof a === 'object')
    .map((a) => {
      const attr: MetadataAttribute = {
        trait_type: asString(a.trait_type, ''),
        value: parseAttributeValue(a.value),
      };
      if (a.display_type != null) {
        attr.display_type = String(a.display_type) as MetadataAttributeDisplayType;
      }
      if (a.max_value != null) {
        const n = Number(a.max_value);
        if (Number.isFinite(n)) attr.max_value = n;
      }
      return attr;
    });
}

function parseAttributeValue(v: unknown): string | number | boolean {
  if (typeof v === 'number' || typeof v === 'boolean') return v;
  return v != null ? String(v) : '';
}

function parseProperties(v: unknown): MetadataProperties | undefined {
  if (v == null || typeof v !== 'object') return undefined;
  const p = v as Record<string, unknown>;

  const props: MetadataProperties = {};

  if (Array.isArray(p.files)) {
    props.files = p.files
      .filter((f): f is Record<string, unknown> => f != null && typeof f === 'object')
      .map((f) => {
        const file: MetadataFile = {
          uri: asString(f.uri, ''),
          type: asString(f.type, 'unknown'),
        };
        if (f.cdn != null) file.cdn = Boolean(f.cdn);
        return file;
      });
  }

  if (p.category != null) {
    props.category = String(p.category) as MetadataCategory;
  }

  if (Array.isArray(p.creators)) {
    props.creators = p.creators
      .filter((c): c is Record<string, unknown> => c != null && typeof c === 'object')
      .map((c) => ({
        address: asString(c.address, ''),
        share: Number.isFinite(Number(c.share)) ? Number(c.share) : 0,
      }));
  }

  return props;
}

function parseCollection(v: unknown): MetadataCollection | undefined {
  if (v == null || typeof v !== 'object') return undefined;
  const c = v as Record<string, unknown>;

  const col: MetadataCollection = {};
  if (c.name != null) col.name = String(c.name);
  if (c.family != null) col.family = String(c.family);

  return col.name != null || col.family != null ? col : undefined;
}
